cpp/src/parquet/encoding.cc

```diff
@@ -2466,6 +2468,99 @@ class DeltaByteArrayDecoder : public DecoderImpl,
 // ----------------------------------------------------------------------
 // BYTE_STREAM_SPLIT
 
+namespace {
+template <typename T, bool has_sse2_or_higher>
+struct ByteStreamSplitDecodePath {
+  static int Decode(T* buffer, int max_values, int num_values_in_buffer,
+                    const uint8_t* data, int& num_values_decoded, int& len_remaining) {
+    throw ParquetException("This should be unreachable.");
+  }
+};
+
+template <typename T>
+struct ByteStreamSplitDecodePath<T, true> {
+  static int Decode(T* buffer, int max_values, int num_values_in_buffer,
+                    const uint8_t* data, int& num_values_decoded, int& len_remaining) {
+    constexpr size_t num_streams = sizeof(T);
+    static_assert(num_streams == 4U || num_streams == 8U);
+    constexpr size_t num_streams_log2 = (num_streams == 8U ? 3U : 2U);
+
+    const int values_to_decode = std::min(num_values_decoded, max_values);
+    const int num_decoded_previously = num_values_in_buffer - num_values_decoded;
+    size_t size = values_to_decode * sizeof(T);
+    const size_t block_size = sizeof(__m128i) * num_streams;
+    const size_t num_blocks = size / block_size;
+    uint8_t* output_data = reinterpret_cast<uint8_t*>(buffer);
+
+    // First handle suffix.
+    // This helps catch if the simd-based processing overflows into the suffix
+    // since almost surely a test would fail.
+    const int num_processed_elements = (num_blocks * block_size) / num_streams;
+    for (int i = num_processed_elements; i < values_to_decode; ++i) {
+      uint8_t gathered_byte_data[num_streams];
+      for (size_t b = 0; b < num_streams; ++b) {
+        const size_t byte_index = b * num_values_in_buffer + num_decoded_previously + i;
+        gathered_byte_data[b] = data[byte_index];
+      }
+      buffer[i] = arrow::util::SafeLoadAs<T>(&gathered_byte_data[0]);
+    }
```

Comment:
instead of copying the code from below, maybe just call ByteStreamSplitDecodePath<T, false> here?

Reply:
This might not be that nice since it would require some changes to the arguments being passed. I think the loop is very small as it is.